### ISSUE TITLE
Custom clusters do not support endpoint discovery through EDS

### ISSUE THREAD

**Issue by pianiststickman** [envoyproxy/envoy#15663]
*Title*: Custom clusters do not support endpoint discovery through EDS

*Description*:
Usage of eds_cluster_config is restricted to clusters that have type explicitly set to `EDS`. Envoy also uses this type for determining whether clusters are initialized during the primary or secondary phase, meaning it is not possible to write a custom cluster type that performs endpoint discovery via EDS.

Instead, I think the custom cluster type should specify whether it utilizes EDS for endpoint discovery, and if so, Envoy should treat it the same way as it does the native EDS cluster (i.e. the cluster is permitted to use `eds_cluster_config`, it is initialized during secondary phase if the config source is not file-based, etc.).

**Comment by lizan** [comment]
Yes this is a valid point, we didn't expose that because no one raised it before. It shouldn't be hard to implement.
